client/state/sharing/keyring/selectors.js

```diff
@@ -49,6 +56,50 @@ export function getUserConnections( state, userId ) {
 }
 
 /**
+ * Given a service, returns a flattened array of all possible accounts for the
+ * service for which a connection can be created.
+ *
+ * @param  {Object} state   Global state tree
+ * @param  {String} service The name of the service to check
+ * @return {Array}          Flattened array of all possible accounts for the service
+ */
+export function getAvailableExternalConnections( state, service ) {
+	const siteUserConnectionsForService = getSiteUserConnectionsForService(
+		state, getSelectedSiteId( state ), getCurrentUserId( state ), service
+	);
+
+	// Iterate over Keyring connections for this service and generate a
+	// flattened array of all accounts, including external users
+	return flatten( getKeyringConnectionsByName( state, service ).map( ( keyringConnection ) => {
+		const accounts = [ {
+			name: keyringConnection.external_display || keyringConnection.external_name,
+			picture: keyringConnection.external_profile_picture,
+			keyringConnectionId: keyringConnection.ID,
+			isConnected: some( siteUserConnectionsForService, {
+				keyring_connection_ID: keyringConnection.ID,
+				external_ID: keyringConnection.external_ID,
+			} ),
+		} ];
+
+		keyringConnection.additional_external_users.forEach( ( externalUser ) => {
```

Comment:
>is `additional_external_users` always guaranteed to be an array?

The logic is copied from [existing code](https://github.com/Automattic/wp-calypso/blob/0798dd93c73c39aba41c278b6fe8a9e289215a23/client/my-sites/sharing/connections/service-connections.js#L297-L336) so I'd expect it to work just as well.